### PR TITLE
Add FXIOS-7652 [v122] Implemented getAddress and addAddress methods, along with corresponding test cases for each

### PR DESCRIPTION
**Pull request by jnrahme**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-7652)

To implement the address autofill feature, we need to retrieve a user's stored addresses. Therefore, I have implemented three functions: getAddress, addAddress, and listAllAddresses to fulfill this functionality and test it. 

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [x] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed I updated documentation / comments for complex code and public methods

### CODE REVIEW

**Comment by lmarceau** [comment]
Hello! Could you make sure that the GitHub ticket is linked (it's part of the template) and fill in the description of the changes? 🙏

**Comment by yoanarios** [comment]
> Hello! Could you make sure that the GitHub ticket is linked (it's part of the template) and fill in the description of the changes? 🙏

Also please fill in the description with a quick note of what was done in the PR so the reviewer can have a better context

**Review by yoanarios — COMMENTED** [review]
@jnrahme LGTM but I don't have much context on this work, I'm wondering if someone else with more context should review it instead

File: Storage/Rust/RustAutofill.swift [hunk 1]
```diff
@@ -202,12 +280,102 @@ public class RustAutofill {
...
+    ///
+    /// - Parameters:
+    ///   - address: UpdatableAddressFields representing the address to be added.
+    ///   - completion: A closure called upon completion with the added address or an error.
+    public func addAddress(address: UpdatableAddressFields, completion: @escaping (Address?, Error?) -> Void) {
+        performDatabaseOperation { error in
+            if let error = error {
+                completion(nil, error)
+                return
+            }
```
Comment (nbhasin2) [inline Storage/Rust/RustAutofill.swift:293]: nit: guard if possible

File: Storage/Rust/RustAutofill.swift [hunk 2]
```diff
@@ -202,12 +280,102 @@ public class RustAutofill {
...
+                } else {
+                    // Handle the case where addAddress returns nil, possibly due to an internal error.
+                    let internalError = NSError(domain: "YourDomain", code: 1, userInfo: [
+                        NSLocalizedDescriptionKey: "Internal error: Failed to add address."
+                    ])
+                    completion(nil, internalError)
+                }
+            } catch let err as NSError {
+                // Handle any other errors that might occur during the database operation.
+                completion(nil, err)
```
Comment (nbhasin2) [inline Storage/Rust/RustAutofill.swift:308]: do you want to log error here? you are catching it anyways

File: Storage/Rust/RustAutofill.swift [hunk 3]
```diff
@@ -202,12 +280,102 @@ public class RustAutofill {
...
+    ///
+    /// - Parameters:
+    ///   - id: The identifier of the address.
+    ///   - completion: A closure called upon completion with the retrieved address or an error.
+    public func getAddress(id: String, completion: @escaping (Address?, Error?) -> Void) {
+        performDatabaseOperation { error in
+            if let error = error {
+                completion(nil, error)
+                return
+            }
```
Comment (nbhasin2) [inline Storage/Rust/RustAutofill.swift:323]: nit: guard if possible

File: firefox-ios/Tests/StorageTests/RustAutofillTests.swift [hunk 4]
```diff
@@ -44,6 +44,79 @@ class RustAutofillTests: XCTestCase {
...
+                XCTAssertEqual(address.guid, retrievedAddress.guid)
+                expectationGetAddress.fulfill()
+            }
+        }
+
+        waitForExpectations(timeout: 3, handler: nil)
+    }
+
+    func testListAllAddressesSuccess() {
+        let expectationListAddresses = expectation(description: "Completes the list all addresses operation")
```
Comment (nbhasin2) [inline firefox-ios/Tests/StorageTests/RustAutofillTests.swift:99]: what about a test case for empty or no addresses?

**Comment by nbhasin2** [comment]
similar nits @jnrahme

File: firefox-ios/Tests/StorageTests/RustAutofillTests.swift (hunk 4 above)
Comment (lmarceau) [reply firefox-ios/Tests/StorageTests/RustAutofillTests.swift:99]: @jnrahme which other test? Cause we have different tests "styles" in the codebase. Using mocks is normally our preferred standard with protocols as you said

File: Storage/Rust/RustAutofill.swift [hunk 5]
```diff
@@ -202,12 +280,102 @@ public class RustAutofill {
         }
     }
 
+    /// Adds an address to the database.
+    ///
+    /// - Parameters:
+    ///   - address: UpdatableAddressFields representing the address to be added.
+    ///   - completion: A closure called upon completion with the added address or an error.
+    public func addAddress(address: UpdatableAddressFields, completion: @escaping (Address?, Error?) -> Void) {
```
Comment (cyndichin) [inline Storage/Rust/RustAutofill.swift:288]: curious about the completion param types. It seems that we want to either returning an Address or an Error, but never a case of `completion(nil, nil)` or `completion(address, error)`. However, it seems we are allowing those possibilities, so I'm curious  to understand why did we decide to go through using `(Address?, Error?)` as opposed to using a Result type in which we specify a `.failure(error)` or a `.success(address)` case. 

happy to chat live to better understand :) thanks!

File: Storage/Rust/RustAutofill.swift [hunk 6]
```diff
@@ -172,6 +240,12 @@ public class RustAutofill {
         }
     }
 
+    /// Marks a credit card as used in the database.
```
Comment (cyndichin) [inline Storage/Rust/RustAutofill.swift:243]: thank you for these comments 🔥

File: Storage/Rust/RustAutofill.swift (hunk 5 above)
Comment (nbhasin2) [reply Storage/Rust/RustAutofill.swift:288]: @cyndichin Good catch!

Blame it on the legacy code refactor but if @jnrahme wants to refactor it in favour of the using proper convention then I would much likely prefer that :)

Comment (cyndichin) [reply Storage/Rust/RustAutofill.swift:288]: ah ok great, thanks for the clarity nish!

File: Storage/Rust/RustAutofill.swift (hunk 2 above)
Comment (jnrahme) [reply Storage/Rust/RustAutofill.swift:307]: I followed the same approach as the credit card code was written. It seems that we log things once the completion block returns an error in the function that calls the function.

File: Storage/Rust/RustAutofill.swift (hunk 5 above)
Comment (jnrahme) [reply Storage/Rust/RustAutofill.swift:287]: https://mozilla-hub.atlassian.net/browse/FXIOS-7810 Ticket created

File: firefox-ios/Tests/StorageTests/RustAutofillTests.swift (hunk 4 above)
Comment (jnrahme) [reply firefox-ios/Tests/StorageTests/RustAutofillTests.swift]: In RustAutoFill, there are currently no protocols available. To establish an abstraction layer for creating a clean mock, I might need to resort to using inheritance and mocking the entire functionality. However, for the time being, testing the empty state seems to be the most appropriate approach.

Comment (jnrahme) [reply firefox-ios/Tests/StorageTests/RustAutofillTests.swift]: @lmarceau, if you examine the existing credit card tests, you'll notice that I followed the same approach used in those tests.

File: Storage/Rust/RustAutofill.swift (hunk 6 above)
Comment (jnrahme) [reply Storage/Rust/RustAutofill.swift:242]: Gracias :)

**Review by nbhasin2 — APPROVED** [review]

**Pull request merged**